Cannot adapt method error when upgrading to Micronaut 3.0.0-M2

During the upgrade of Micronaut Flyway to Micronaut 3.0.0-M2 so test fail with:

```
  Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [io.micronaut.flyway.EventListenerSpec$TestEventListener$ApplicationEventListener$onStartupCleaned2$Intercepted]: Cannot adapt method [void onApplicationEvent(Object event)]. Target method [onStartupCleaned] not found on bean class io.micronaut.flyway.EventListenerSpec$TestEventListener
      at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2315)
      at io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3619)
      at io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:3495)
      at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:850)
      at io.micronaut.context.DefaultBeanContext.publishEvent(DefaultBeanContext.java:1662)
      at io.micronaut.flyway.AbstractFlywayMigration.runFlyway(AbstractFlywayMigration.java:95)
      at io.micronaut.flyway.AbstractFlywayMigration.forceRun(AbstractFlywayMigration.java:85)
      at io.micronaut.flyway.AbstractFlywayMigration.run(AbstractFlywayMigration.java:65)
      at io.micronaut.flyway.DataSourceMigrationRunner.lambda$null$0(DataSourceMigrationRunner.java:65)
      at java.util.Optional.ifPresent(Optional.java:159)
      at io.micronaut.flyway.DataSourceMigrationRunner.lambda$onCreated$1(DataSourceMigrationRunner.java:63)
      at java.util.Optional.ifPresent(Optional.java:159)
      at io.micronaut.flyway.DataSourceMigrationRunner.onCreated(DataSourceMigrationRunner.java:60)
      at io.micronaut.flyway.DataSourceMigrationRunner.onCreated(DataSourceMigrationRunner.java:37)
      at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2343)
      at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3226)
      at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3212)
      at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:2603)
      at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1882)
      ... 8 more
  Caused by: java.lang.IllegalStateException: Cannot adapt method [void onApplicationEvent(Object event)]. Target method [onStartupCleaned] not found on bean class io.micronaut.flyway.EventListenerSpec$TestEventListener
      at io.micronaut.aop.chain.AdapterIntroduction.lambda$new$1(AdapterIntroduction.java:75)
      at java.util.Optional.orElseThrow(Optional.java:290)
      at io.micronaut.aop.chain.AdapterIntroduction.<init>(AdapterIntroduction.java:75)
      at io.micronaut.aop.chain.DefaultInterceptorRegistry.resolveInterceptors(DefaultInterceptorRegistry.java:65)
      at io.micronaut.aop.chain.InterceptorChain.resolveInterceptors(InterceptorChain.java:206)
      at io.micronaut.aop.chain.InterceptorChain.resolveIntroductionInterceptors(InterceptorChain.java:144)
      at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2286)
      ... 26 more
```

The test looks like:

```groovy
    @Singleton
    @Requires(property = 'spec.name', value = 'EventListenerSpec')
    static class TestEventListener {
        List<MigrationFinishedEvent> migrationFinishedEvents = []
        List<SchemaCleanedEvent> schemaCleanedEvents = []

        @EventListener
        void onStartupMigrate(MigrationFinishedEvent event) {
            migrationFinishedEvents.add(event)
        }

        @EventListener
        void onStartupCleaned(SchemaCleanedEvent event) {
            schemaCleanedEvents.add(event)
        }
    }
```

## Reproduce the issue

Run the tests in this PR https://github.com/micronaut-projects/micronaut-flyway/pull/212

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,4 +1,4 @@
-projectVersion=3.0.1-SNAPSHOT
+projectVersion=3.0.0-SNAPSHOT
 projectGroupId=io.micronaut
 title=Micronaut
 projectDesc=Natively Cloud Native

--- a/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/InjectVisitor.groovy
+++ b/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/InjectVisitor.groovy
@@ -1413,7 +1413,7 @@ final class InjectVisitor extends ClassCodeVisitorSupport {
                         true,
                         false,
                         originatingElement,
-                        methodAnnotationMetadata,
+                        new AnnotationMetadataHierarchy(concreteClassAnnotationMetadata, methodAnnotationMetadata),
                         [elementFactory.newClassElement(typeToImplement, AnnotationMetadata.EMPTY_METADATA)] as ClassElement[],
                         groovyVisitorContext,
                         configurationMetadataBuilder,

--- a/inject-groovy/src/test/groovy/io/micronaut/aop/adapter/MethodAdapterSpec.groovy
+++ b/inject-groovy/src/test/groovy/io/micronaut/aop/adapter/MethodAdapterSpec.groovy
@@ -22,6 +22,44 @@ import io.micronaut.core.reflect.ReflectionUtils
 import io.micronaut.inject.BeanDefinition
 
 class MethodAdapterSpec extends AbstractBeanDefinitionSpec {
+    void 'test method adapter with failing requirements is not present'() {
+        given:
+        def context = buildContext('''
+package issue5640;
+
+import io.micronaut.aop.Adapter;
+import java.lang.annotation.*;
+import io.micronaut.context.annotation.Requires;
+import static java.lang.annotation.ElementType.*;
+import static java.lang.annotation.RetentionPolicy.*;
+import jakarta.inject.Singleton;
+import static java.nio.charset.StandardCharsets.US_ASCII;
+
+@Singleton
+@Requires(property="not.present")
+class AsciiParser {
+    @Parse
+    public String parseAsAscii(byte[] value) {
+        return new String(value, US_ASCII);
+    }
+}
+
+@Retention(RUNTIME)
+@Target(METHOD)
+@Adapter(Parser.class)
+@interface Parse {}
+
+interface Parser {
+    String parse(byte[] value);
+}
+''')
+        def adaptedType = context.classLoader.loadClass('issue5640.Parser')
+
+        expect:
+        !context.containsBean(adaptedType)
+        context.getBeansOfType(adaptedType).isEmpty()
+    }
+
     void "test method adapter with overloading"() {
         given:
         def context = buildContext('adapteroverloading.Test', '''

--- a/inject-java/src/test/groovy/io/micronaut/aop/adapter/MethodAdapterSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/aop/adapter/MethodAdapterSpec.groovy
@@ -31,6 +31,44 @@ import java.nio.charset.StandardCharsets
 
 class MethodAdapterSpec extends AbstractTypeElementSpec {
 
+    void 'test method adapter with failing requirements is not present'() {
+        given:
+        def context = buildContext('''
+package issue5640;
+
+import io.micronaut.aop.Adapter;
+import java.lang.annotation.*;
+import io.micronaut.context.annotation.Requires;
+import static java.lang.annotation.ElementType.*;
+import static java.lang.annotation.RetentionPolicy.*;
+import jakarta.inject.Singleton;
+import static java.nio.charset.StandardCharsets.US_ASCII;
+
+@Singleton
+@Requires(property="not.present")
+class AsciiParser {
+    @Parse
+    public String parseAsAscii(byte[] value) {
+        return new String(value, US_ASCII);
+    }
+}
+
+@Retention(RUNTIME)
+@Target({ANNOTATION_TYPE, METHOD})
+@Adapter(Parser.class)
+@interface Parse {}
+
+interface Parser {
+    String parse(byte[] value);
+}
+''')
+        def adaptedType = context.classLoader.loadClass('issue5640.Parser')
+
+        expect:
+        !context.containsBean(adaptedType)
+        context.getBeansOfType(adaptedType).isEmpty()
+    }
+
     void 'test method adapter with byte[] argument'() {
         given:
         def context = buildContext('issue5054.AsciiParser', '''

